src/util/circular_buffer.h

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright Redis Ltd. 2016 - present
+ * Licensed under your choice of the Redis Source Available License 2.0 (RSALv2) or
+ * the Server Side Public License v1 (SSPLv1).
+ */
+
+#pragma once
+
+#include <stdbool.h>
+#include <sys/types.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+// forward declaration
+typedef struct _CircularBuffer _CircularBuffer;
+typedef _CircularBuffer* CircularBuffer;
+
+// Creates a new circular buffer, with `cap` items of size `item_size`
+CircularBuffer CircularBuffer_New(size_t item_size, uint cap);
+
+// Returns the number of items in the buffer
+uint64_t CircularBuffer_ItemCount(CircularBuffer cb);
+
+// Returns buffer capacity
+uint64_t CircularBuffer_Cap(CircularBuffer cb);
+
+// Returns the size of each item in the buffer
+uint CircularBuffer_ItemSize(const CircularBuffer cb);
+
+// Returns true if buffer is empty
+bool CircularBuffer_Empty(const CircularBuffer cb);
+
+// Returns true if buffer is full
+bool CircularBuffer_Full(const CircularBuffer cb);
+
+// Adds an item to buffer.
+// Returns 1 on success, 0 otherwise
+// This function is thread-safe and lock-free
+int CircularBuffer_Add(CircularBuffer cb, void *item);
```

Comment:
The purpose of this container is eventually to support holding queries, or in general objects that were allocated on the heap.

It seems like writing can override existing items without destroying the "object".
Don't want the container to have to deal with object destruction.
But maybe provide an overload where the caller can provide a callback to release the existing slot before you write to it.
This also requires you to know if the slot read/write state.

Reply:
Right.
We may add this functionality later on if needed.